independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,117 @@
+package io.quarkus.bootstrap.runner;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.jar.JarFile;
+
+import io.smallrye.common.io.jar.JarFiles;
+
+public class JarFileReference {
+    // Guarded by an atomic reader counter that emulate the behaviour of a read/write lock.
+    // To enable virtual threads compatibility and avoid pinning it is not possible to use an explicit read/write lock
+    // because the jarFile access may happen inside a native call (for example triggered by the RunnerClassLoader)
+    // and then it is necessary to avoid blocking on it.
+    private final JarFile jarFile;
+
+    private final AtomicReference<JarFileReference> jarFileReference;
+
+    private final AtomicInteger referenceCounter = new AtomicInteger(1);
+
+    private JarFileReference(AtomicReference<JarFileReference> jarFileReference, JarFile jarFile) {
+        this.jarFileReference = jarFileReference;
+        this.jarFile = jarFile;
+    }
+
+    /**
+     * Increase the readers counter of the jarFile and returns it.
+     *
+     * @return The jarFile wrapped by this reference or null if a closing has been requested thus making the file no longer
+     *         accessible.
+     */
+    private JarFile acquire() {
+        while (true) {
+            int count = referenceCounter.get();
+            if (count == 0) {
+                return null;
+            }
+            if (referenceCounter.compareAndSet(count, count+1)) {
+                return jarFile;
+            }
+        }
+    }
+
+    /**
+     * Decrease the readers counter of the jarFile.
+     * If the counter drops to 0 and a release has been requested also closes the jarFile.
+     */
+    private void release() {
+        int count = referenceCounter.decrementAndGet();
+        if (count < 0) {
+            throw new IllegalStateException("The reference counter cannot be negative, found: " + count);
+        }
+        if (count == 0) {
+            try {
+                jarFile.close();
```

Comment:
I would pretend the compare and set before closing this and don't only if the previous value was 1

Reply:
Done.

Reply:
I would use compare and set because 2 close can make the counter to go less then 0 breaking the acquire which return null

Reply:
I change this, but please give another look, since I'm not 100% sure and feel free to ping me in case you have any question or doubt.